### PR TITLE
Fix reservations [BBS-221]

### PR DESCRIPTION
**Pull request by tripox**

### CODE REVIEW

File: modules/aleph/src/Aleph/Entity/AlephHoldGroup.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+   * @param \SimpleXMLElement $xml
+   *
+   * @return \Drupal\aleph\Aleph\Entity\AlephHoldGroup
+   */
+  public static function createHoldGroupFromXML(\SimpleXMLElement $xml) {
+    $hold_group = new self();
+    $pickup_locations = [];
+
+    foreach ($xml->xpath('pickup-locations/pickup-location') as $pickup_location) {
+      $pickup_locations[(string) $pickup_location['code']] = (string) $pickup_location;
```
Comment (danquah) [inline modules/aleph/src/Aleph/Entity/AlephHoldGroup.php]: Be carefull here - you havn't cheked that ['code'] actually exists.

File: modules/aleph/includes/aleph.reservation.inc [hunk 2]
```diff
@@ -240,10 +242,13 @@ function aleph_reservation_list($account, $requested_type = NULL) {
  *   Reservation exists.
  * @throws \DingProviderUserException
  *    General exception.
+ * @throws \DingProviderAuthException
```
Comment (danquah) [inline modules/aleph/includes/aleph.reservation.inc]: When is this thrown? (aka, the comment is missing)

File: modules/aleph/src/Aleph/AlephClient.php [hunk 3]
```diff
@@ -308,20 +313,22 @@ public function createReservation(AlephPatron $patron, AlephRequest $request) {
...
-    // BIB library code + the system number.
-    // For example, USM01000050362.
     $rid = $this->mainLibrary . $request->getDocNumber();
 
-    // ADM library code + the item record key.
-    // For example, USM50000238843000320.
-    $iid = $request->getInstitutionCode() . $request->getDocNumber() .
-      $request->getItemSequence();
+    foreach ($holding_groups as $url => $holding_group) {
+      $response = $this->requestRest(
```
Comment (danquah) [inline modules/aleph/src/Aleph/AlephClient.php]: How are errors handled here? RuntimeExceptions? Seems a bit unreliable.
Would be nicer if the function handled the errors and had a clear documented behaviour when errors orccured such as returning NULL or a business-exception.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 4]
```diff
@@ -386,4 +393,21 @@ public function getPatronBlocks($bor_id) {
...
+  /**
+   * @param \Drupal\aleph\Aleph\Entity\AlephPatron $patron
+   *    The Aleph patron.
+   * @param \Drupal\aleph\Aleph\Entity\AlephMaterial $material
+   *    The Aleph material.
+   *
+   * @return \SimpleXMLElement[]
+   * @throws \RuntimeException
+   */
+  public function getHoldingGroups(AlephPatron $patron, AlephMaterial $material) {
```
Comment (danquah) [inline modules/aleph/src/Aleph/AlephClient.php]: Same as above, consider a more robust error-handling.

File: modules/aleph/src/Aleph/Entity/AlephRequest.php [hunk 5]
```diff
@@ -124,6 +124,15 @@ class AlephRequest {
    */
   protected $institutionCode;
 
+  /**
+   * Array of URLs to hold groups.
+   *
+   * @var string[]
+   * For example:
+   * http://<server>/rest-dlf/patron/<patron_id>/record/ICE01000335901/holds/ICE53000335901001570
```
Comment (danquah) [inline modules/aleph/src/Aleph/Entity/AlephRequest.php:132]: :+1:

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 6]
```diff
@@ -255,6 +262,22 @@ public function deleteReservation($patron, $reservation) {
     return AlephRequestResponse::createRequestResponseFromXML($response);
   }
 
+  /**
+   * @param AlephPatron $patron
+   * @param AlephMaterial $material
+   *
+   * @return AlephHoldGroup[]
+   * @throws \RuntimeException
```
Comment (danquah) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: Throwing runtime-exceptions is a bit like saying "I might have messed up, and now it's your problem". Use a business-exception instead (eg. a Aleph exception) or handle the issue and return something well-defined such ash NULL, FALSE or []

**Review by danquah — CHANGES_REQUESTED** [review]
Would love some better error-handling.

File: modules/aleph/aleph.module [hunk 7]
```diff
@@ -222,6 +223,15 @@ function aleph_form_profile2_form_alter(&$form, &$form_state) {
   $form['profile_provider_aleph']['field_aleph_expiry_date']['#disabled'] = TRUE;
 }
 
+/**
+ * Disable the 'update reservation' button as it's not possible in Aleph.
```
Comment (kasperg) [inline modules/aleph/aleph.module]: General style issue: Consider changing the docblock to `implements hook_form_alter()`.

In my opinion the specifics about what the hook does belong in the inline comments - not in the general DocBlock. If more functionality is added to the form alter then the DocBlock comment quickly becomes obsolete.

You might also consider `hook_form_form_id_alter()`.

File: modules/aleph/includes/aleph.reservation.inc [hunk 8]
```diff
@@ -146,16 +149,21 @@ function aleph_reservation_default_interest_period($profile = NULL) {
...
  */
 function aleph_reservation_list($account, $requested_type = NULL) {
   // Throw exception if we're not logged in.
-  ding_user_get_creds($account);
+  try {
+    ding_user_get_creds();
+  }
+  catch (DingProviderAuthException $e) {
+    return [];
+  }
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: Why is this change needed?

As I understand it we want the `DingProviderAuthException` to be thrown because it signals to the calling code that an authentication dialog should be shown.

File: modules/aleph/includes/aleph.reservation.inc [hunk 9]
```diff
@@ -258,40 +273,50 @@ function aleph_reservation_create($account, $id) {
   $interest_period = aleph_reservation_default_interest_period();
 
   $end_request_date = new DateTime();
-  $end_request_date->add(new DateInterval('P' . $interest_period . 'D'));
+  try {
+    $end_request_date->add(new DateInterval('P' . $interest_period . 'D'));
+  }
+  catch (Exception $e) {
+  }
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: What is going on here?

Two code smells:
- Catching the generic exception. This may hide all sorts of problems as more and more PHP functions throw exceptions instead of raising errors, warnings etc.
- Catching an exception and doing nothing. Exceptions are usually the result of a problem and it usually warrants a comment about why no action is taken.

File: modules/aleph/includes/aleph.reservation.inc [hunk 10]
```diff
@@ -258,40 +273,50 @@ function aleph_reservation_create($account, $id) {
...
-  $patron = $patron_handler->getPatron();
   $response = NULL;
   try {
-    $response = $patron_handler->createReservation($patron, $reservation);
+    $response = $patron_handler->createReservation($patron, $reservation, $holding_groups);
+    if ($response->success()) {
+      return array(
+        'branch' => $request->getPickupLocation(),
+        // Queue number not supported by Aleph.
+        'queue_number' => NULL,
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc:302]: :+1:

File: modules/aleph/includes/aleph.reservation.inc [hunk 11]
```diff
@@ -302,10 +327,12 @@ function aleph_reservation_update() {
...
  *
- * @throws \DingProviderAuthException
- * @throws \DingProviderUserException
- *
  * @return bool
+ *   If the deletion succeeded or not.
+ *
+ * @throws DingProviderAuthException
+ * @throws DingProviderReservationUserError
+ * @throws \Drupal\aleph\Aleph\Exception\AlephClientException
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: Either use the fully qualified class names or all classes or for none.

File: modules/aleph/includes/aleph.reservation.inc [hunk 12]
```diff
@@ -368,30 +395,19 @@ function aleph_reservation_branch_name($branch_id) {
...
  */
 function aleph_reservation_pickup_branches($account) {
   // Throw exception if we're not logged in.
-  ding_user_get_creds($account);
+  try {
+    ding_user_get_creds();
+  }
+  catch (DingProviderAuthException $e) {
+    return [];
+  }
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: See comment above about `DingProviderAuthException`.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 13]
```diff
@@ -247,13 +249,13 @@ public function getItems(AlephMaterial $material) {
    * @param \Drupal\aleph\Aleph\Entity\AlephPatron $patron
    *    The patron to get loans from.
    *
-   * @param $loan_id
+   * @param bool $loan_id
    *    The loan ID to get specific loan.
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: `bool`? That seems odd compared to the description of the parameter.

File: modules/aleph/includes/aleph.reservation.inc (hunk 8 above)
Comment (kasperg) [reply modules/aleph/includes/aleph.reservation.inc]: Also: If this change is needed then the code comment above seems obsolete.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 14]
```diff
@@ -290,16 +292,22 @@ public function getReservations(AlephPatron $patron) {
...
+   *   The holding groups.
    *
    * @return \SimpleXMLElement
-   * @throws \RuntimeException
+   *   The SimpleXMLElement from the raw XML response.
+   *
+   * @throws AlephClientException
    */
-  public function createReservation(AlephPatron $patron, AlephRequest $request) {
+  public function createReservation(AlephPatron $patron, AlephRequest $request, $holding_groups) {
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: Please add `array` as a type hint for `$holding_groups`.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 15]
```diff
@@ -290,16 +292,22 @@ public function getReservations(AlephPatron $patron) {
...
     $options = array();
+    $response = FALSE;
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: Why do we need to set response to `FALSE`? It does not seem to be in line with the parameter declaration.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 16]
```diff
@@ -308,20 +316,22 @@ public function createReservation(AlephPatron $patron, AlephRequest $request) {
...
 
-    // BIB library code + the system number.
-    // For example, USM01000050362.
     $rid = $this->mainLibrary . $request->getDocNumber();
 
-    // ADM library code + the item record key.
-    // For example, USM50000238843000320.
-    $iid = $request->getInstitutionCode() . $request->getDocNumber() .
-      $request->getItemSequence();
+    foreach ($holding_groups as $url => $holding_group) {
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: A comment about what is going on here would probably be good.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 17]
```diff
@@ -380,10 +390,27 @@ public function deleteReservation(AlephPatron $patron, AlephRequest
...
+  /**
+   * @param \Drupal\aleph\Aleph\Entity\AlephPatron $patron
+   *    The Aleph patron.
+   * @param \Drupal\aleph\Aleph\Entity\AlephMaterial $material
+   *    The Aleph material.
+   *
+   * @return \SimpleXMLElement[]
+   * @throws AlephClientException
+   */
+  public function getHoldingGroups(AlephPatron $patron, AlephMaterial $material) {
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: Please add a comment about what this function does.

The concept of Holding Groups is unique to Aleph.

File: modules/aleph/src/Aleph/Entity/AlephHoldGroup.php [hunk 18]
```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace Drupal\aleph\Aleph\Entity;
+
+/**
+ * Class AlephHoldGroup.
+ *
+ * Entity for hold groups.
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Entity/AlephHoldGroup.php:8]: From a documentation perspective the DocBlocks in this class adds close to nothing.

File: modules/aleph/src/Aleph/Entity/AlephHoldGroup.php [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+   * @return mixed
+   */
+  public function getSubLibrary() {
+    return $this->subLibrary;
+  }
+
+  /**
+   * @param mixed $subLibrary
+   */
+  public function setSubLibrary($subLibrary) {
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Entity/AlephHoldGroup.php]: Could the visibility of the setters here be reduced to `protected`? Once we have a holding group I do not think we want outsiders to change the properties.

`protected` or even `private` may be usable within our factory method.

File: modules/aleph/src/Aleph/Exceptions/AlephClientException.php [hunk 20]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Drupal\aleph\Aleph\Exception;
+
+/**
+ * Define exceptions for different error conditions inside the Aleph client.
+ */
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Exceptions/AlephClientException.php]: This DocBlock seems odd.

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 21]
```diff
@@ -105,6 +107,9 @@ public function getLoans() {
    *
    * @return bool
    *    True if setting new pincode succeeded.
+   *
+   * @throws AlephClientException
+   * @throws \AlephPatronInvalidPin
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: `\AlephPatronInvalidPin` is no longer in the global namespace.

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 22]
```diff
@@ -225,17 +230,22 @@ public function renewLoans($ids) {
   /**
    * Create a reservation for a patron.
    *
-   * @param AlephPatron $patron
+   * @param AlephPatron      $patron
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: Odd spacing?

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 23]
```diff
@@ -255,6 +265,22 @@ public function deleteReservation($patron, $reservation) {
...
+   * @return AlephHoldGroup[]
+   * @throws AlephClientException
+   */
+  public function getHoldingGroups($patron, $material) {
+    $groups = $this->client->getHoldingGroups($patron, $material);
+    $result = [];
+    foreach ($groups as $group) {
+      $result[(string) $group['href']] = AlephHoldGroup::createHoldGroupFromXML($group);
+    }
+    return $result;
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: 💡 Perfect use case for `array_map`

File: modules/aleph/includes/aleph.reservation.inc [hunk 24]
```diff
@@ -258,40 +273,50 @@ function aleph_reservation_create($account, $id) {
...
+  $material->setId($id);
   $request->setRequestDate(new DateTime());
   $request->setEndRequestDate($end_request_date);
   $request->setPickupLocation($defaults['aleph_preferred_branch']);
+  $request->setDocNumber($id);
   $reservation->setRequest($request);
 
+  try {
+    $holding_groups = $patron_handler->getHoldingGroups($patron, $material);
+  }
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc:289]: This way we retrieve the holding groups from the patron instead of basing them on the configuration on the site. Is that intentional?

**Review by kasperg — CHANGES_REQUESTED** [review]
Reviewed. Lots of comments - big and small.

File: modules/aleph/includes/aleph.reservation.inc (hunk 24 above)
Comment (tripox) [reply modules/aleph/includes/aleph.reservation.inc:289]: The holding groups are different from patron to patron.
It is intentional. :)

File: modules/aleph/aleph.module [hunk 25]
```diff
@@ -222,6 +223,14 @@ function aleph_form_profile2_form_alter(&$form, &$form_state) {
   $form['profile_provider_aleph']['field_aleph_expiry_date']['#disabled'] = TRUE;
 }
 
+/**
+ * Implements hook_form_form_id_alter().
+ */
+function aleph_form_ding_reservation_reservations_notready_form_alter(&$form, &$form_state, $form_id) {
+  // Disable the 'update reservation' button as it's not possible in Aleph.
+  unset($form['actions_container']['actions_top']['update']);
```
Comment (kasperg) [inline modules/aleph/aleph.module:231]: :+1:

File: modules/aleph/includes/aleph.reservation.inc [hunk 26]
```diff
@@ -146,16 +149,16 @@ function aleph_reservation_default_interest_period($profile = NULL) {
...
+ *   Type of reservations to list. NULL for all.
  *
  * @return \DingProviderReservation[]
- *    Array of DingProviderReservations.
+ *   Array of DingProviderReservations.
  */
 function aleph_reservation_list($account, $requested_type = NULL) {
   // Throw exception if we're not logged in.
-  ding_user_get_creds($account);
+  ding_user_get_creds();
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc:161]: Is there a reason why we are not passing the `$accout` parameter anymore?

File: modules/aleph/includes/aleph.reservation.inc (hunk 24 above)
Comment (kasperg) [reply modules/aleph/includes/aleph.reservation.inc:289]: OK. I just worried if we were going to be creating reservations for items outside the scope of the site.

File: modules/aleph/includes/aleph.reservation.inc [hunk 27]
```diff
@@ -368,30 +390,17 @@ function aleph_reservation_branch_name($branch_id) {
...
- *    The branches as key and name.
+ *   The branches as key and name.
+ *
+ * @throws DingProviderAuthException
+ *    If the patron is not logged in.
  */
 function aleph_reservation_pickup_branches($account) {
   // Throw exception if we're not logged in.
-  ding_user_get_creds($account);
+  ding_user_get_creds();
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc:402]: As above.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 28]
```diff
@@ -98,7 +99,7 @@ public function request($method, $operation, array $params = array()) {
     }
 
     // Throw exception if the status from Aleph is not OK.
-    throw new \RuntimeException('Request error: ' . $response->code . $response->error);
+    throw new AlephClientException('Request error: ' . $response->code . $response->error);
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: [The default Exception constructor already have separate arguments for code and error message](http://php.net/manual/en/language.exceptions.extending.php). You may not want to combine them into a single string unless there is a reason to.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 29]
```diff
@@ -380,10 +392,35 @@ public function deleteReservation(AlephPatron $patron, AlephRequest
...
+   * @return \SimpleXMLElement[]
+   *   XML response from Aleph with holding groups.
+   *
+   * @throws AlephClientException
+   */
+  public function getHoldingGroups(AlephPatron $patron, AlephMaterial $material) {
+    return $this->requestRest(
+      'GET',
+      'patron/' . $patron->getId() . '/record/' . $this->mainLibrary .
+      $material->getId() . '/holds?view=full&institution=ICE53'
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: This hardcodes the institution within the client. That does not seem right?

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php (hunk 23 above)
Comment (kasperg) [reply modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: Why is the URL (`$group['href']`) not a property of a holding group?

File: modules/aleph/includes/aleph.reservation.inc (hunk 26 above)
Comment (tripox) [reply modules/aleph/includes/aleph.reservation.inc:161]: It's not used by ding_user_get_creds().

Comment (kasperg) [reply modules/aleph/includes/aleph.reservation.inc:161]: :+1: Gah. Good point!

File: modules/aleph/src/Aleph/AlephClient.php (hunk 28 above)
Comment (kasperg) [reply modules/aleph/src/Aleph/AlephClient.php]: You changed this on L101 but not here?

File: modules/aleph/src/Aleph/Entity/AlephHoldGroup.php [hunk 30]
```diff
@@ -0,0 +1,125 @@
...
+    return $this->url;
+  }
+
+  /**
+   * Set the hold group URL.
+   *
+   * @param string $url
+   *   The hold group URL.
+   */
+  public function setUrl($url) {
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Entity/AlephHoldGroup.php]: This setter is public. The others are protected. Is that intentional?

**Review by kasperg — COMMENTED** [review]
Just minor comments.

**Review by kasperg — APPROVED** [review]
